src/main/java/ch/njol/skript/sections/EffSecShoot.java

```diff
@@ -0,0 +1,281 @@
+			"shoot %entitydatas% [from %livingentities/locations%] [(at|with) (speed|velocity) %-number%] [%-direction%]",
+			"(make|let) %livingentities/locations% shoot %entitydatas% [(at|with) (speed|velocity) %-number%] [%-direction%]"
+		);
+		EventValues.registerEventValue(ShootEvent.class, Entity.class, new Getter<Entity, ShootEvent>() {
+			@Override
+			public @Nullable Entity get(ShootEvent shootEvent) {
+				return shootEvent.getProjectile();
+			}
+		}, EventValues.TIME_NOW);
+		EventValues.registerEventValue(ShootEvent.class, Projectile.class, new Getter<Projectile, ShootEvent>() {
+			@Override
+			public @Nullable Projectile get(ShootEvent shootEvent) {
+				return shootEvent.getProjectile() instanceof Projectile projectile ? projectile : null;
+			}
+		}, EventValues.TIME_NOW);
+
+		if (!Skript.isRunningMinecraft(1, 20, 3)) {
+			try {
+				launchWithBukkitConsumer = LivingEntity.class.getMethod("launchProjectile", Class.class, Vector.class, org.bukkit.util.Consumer.class);
+			} catch (NoSuchMethodException e) {
+				throw new RuntimeException(e);
+			}
+		}
+		boolean launchHasJavaConsumer = Skript.methodExists(LivingEntity.class, "launchProjectile", Class.class, Vector.class, Consumer.class);
+		RUNNING_PAPER = launchWithBukkitConsumer != null || launchHasJavaConsumer;
+	}
+
+	private final static Double DEFAULT_SPEED = 5.;
+	private Expression<EntityData<?>> types;
+	private Expression<?> shooters;
+	private @Nullable Expression<Number> velocity;
+	private @Nullable Expression<Direction> direction;
+	public static Entity lastSpawned = null;
+	private @Nullable Trigger trigger;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult, @Nullable SectionNode sectionNode, @Nullable List<TriggerItem> triggerItems) {
+		//noinspection unchecked
+		types = (Expression<EntityData<?>>) exprs[matchedPattern];
+		shooters = exprs[1 - matchedPattern];
+		//noinspection unchecked
+		velocity = (Expression<Number>) exprs[2];
+		//noinspection unchecked
+		direction = (Expression<Direction>) exprs[3];
+
+		if (sectionNode != null) {
+			AtomicBoolean delayed = new AtomicBoolean(false);
+			Runnable afterLoading = () -> delayed.set(!getParser().getHasDelayBefore().isFalse());
+			trigger = loadCode(sectionNode, "shoot", afterLoading, ShootEvent.class);
+			if (delayed.get()) {
+				Skript.error("Delays cannot be used within a 'shoot' effect section");
+				return false;
+			}
+		}
+		return true;
+	}
+
+	@Override
+	protected @Nullable TriggerItem walk(Event event) {
+		lastSpawned = null;
+		Number finalVelocity = velocity != null ? velocity.getSingle(event) : DEFAULT_SPEED;
+		Direction finalDirection = direction != null ? direction.getSingle(event) : Direction.IDENTITY;
+		if (finalVelocity == null || finalDirection == null)
+			return null;
+
+		for (Object shooter : shooters.getArray(event)) {
+			for (EntityData<?> entityData : types.getArray(event)) {
```

Comment:
need to get entitydatas outside of the loops, otherwise it'll be evaluated multiple times